java/src/jmri/jmrix/dccpp/swing/DCCppRosterExportAction.java

```diff
@@ -0,0 +1,68 @@
+package jmri.jmrix.dccpp.swing;
+
+import java.awt.event.ActionEvent;
+import javax.swing.AbstractAction;
+import javax.swing.JOptionPane;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import javax.script.Bindings;
+import javax.script.ScriptException;
+import javax.script.SimpleBindings;
+
+import jmri.JmriException;
+import jmri.script.ScriptEngineSelector;
+
+/**
+ * Menu action to Export all JMRI Roster Entries as ROSTER() macro calls
+ *  This adds a menu item to the "DCC++" menu. Taking this option opens a 
+ *  window where the ROSTER() lines can be copied into your myAutomation.h 
+ *  Syntax: ROSTER(999,"Loco Name","F0/F1/*F2/F3/F4/F5/F6/F7/F8")
+ * NOTE: calls existing script for compatibility
+ * @author MSteveTodd Copyright (C) 2023
+ */
+public class DCCppRosterExportAction extends AbstractAction {
+
+    private final ScriptEngineSelector _scriptEngineSelector = new ScriptEngineSelector();
+    private final String scriptFilename = "program:jython"+File.separator+"DCC-EX"+File.separator+"RosterExportToDCC-EX.py";
+    private static final Logger log = LoggerFactory.getLogger(DCCppRosterExportAction.class);
+    
+    public DCCppRosterExportAction(String s) {
+        super(s);
+    }
+
+    public DCCppRosterExportAction() {
+        this("Roster Export to DCC-EX");
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        log.debug("in actionPerformed");
+        ScriptEngineSelector.Engine engine =
+                    _scriptEngineSelector.getSelectedEngine();
+        if (engine == null) {
+            String em = "Script engine is null";
+            log.error(em);
+            JOptionPane.showMessageDialog(null,
+                    em, "Script Error",
+                    JOptionPane.ERROR_MESSAGE);
+            return;
+        }
+        try (InputStreamReader reader = new InputStreamReader(
+                    new FileInputStream(jmri.util.FileUtil.getExternalFilename(scriptFilename)),
+                    StandardCharsets.UTF_8)) {
+            engine.getScriptEngine().eval(reader);
+        } catch (IOException | ScriptException ex) {
+            String em = "Cannot execute script:" + ex;
+            log.error(em);
```

Comment:
Static analysis fails on line 60.

`Error:  High: Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance [jmri.jmrix.dccpp.swing.DCCppRosterExportAction] At DCCppRosterExportAction.java:[line 60] SLF4J_FORMAT_SHOULD_BE_CONST`

Either you could suppress this error or you could change line 60 to:
log.error("Cannot execute script: {}", ex);

Reply:
@danielb is there a way to run that test locally?

Reply:
The static analysis CI test is defined in this file:
https://github.com/JMRI/JMRI/blob/master/.github/workflows/static-analysis.yml

I think it's this that you should run, but I'm not sure. I guess only the last line is enough.
```
# compile with ECJ for warnings or errors
mvn antrun:run -Danttarget=tests-warnings-check
# run Spotbugs and Checkstyle
mvn clean test -U -P travis-spotbugs --batch-mode --file=pom.xml
```

Reply:
My goal was to have the log message and the showMessageDialog show the same error without actually duplicating it in the code. I must be missing the mechanism for that. Any suggestions?

Reply:
In this case, your original solution might have been the right one. And that the solution is to suppress the error. The way to do that is this. Example from `apps.AppsBase`, line 61:
```
    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings( value="SLF4J_FORMAT_SHOULD_BE_CONST",
        justification="Info String always needs to be evaluated")
    static public void preInit(String applicationName) {
```

Usually logging statements are unique so it's better to let the logging statement itself build the logging message. But since you show a dialog with the exact same message directly after the logging statement, you have a different case than usual.

So my suggestion is that you revert to the original code:
```
            String em = "Cannot execute script:" + ex;
            log.error(em);
```
and then suppress this error.

I apologize for not giving this advice in the first place, but I wasn't sure on how to suppress it. But a search for the error let me found examples where it's already suppressed.